query-planner-js/src/Scope.ts

```diff
@@ -0,0 +1,259 @@
+    if (directives && directives.length == 0) {
+      directives = undefined;
+    }
+    // If we have directives, we always want to preserve the condition so as to preserve that directive.
+    if (directives) {
+      return new Scope(this.context, type, directives, this);
+    }
+    // Scope always preserve the immediate parent type, but if the new type is already the parent one
+    // and we have not directives to preserve, simply return the existing scope.
+    if (!directives && type === this.parentType) {
+      return this;
+    }
+    return new Scope(this.context, type, directives, Scope.pruneRefinedTypes(this, type));
+  }
+
+  private static pruneRefinedTypes(
+    toPrune: Scope | undefined,
+    refiningType: GraphQLCompositeType
+  ) : Scope | undefined {
+    if (!toPrune) {
+      return undefined;
+    }
+    if (!toPrune.directives && isTypeSubTypeOf(toPrune.context.schema, refiningType, toPrune.parentType)) {
+      // The newly added type is a subtype of the current "link", and the current link has no directives,
+      // so it's not useful anymore. Skip it, and check if we can prune further.
+      return Scope.pruneRefinedTypes(toPrune.enclosing, refiningType);
+    }
+    return new Scope(
+      toPrune.context,
+      toPrune.parentType,
+      toPrune.directives,
+      Scope.pruneRefinedTypes(toPrune.enclosing, refiningType)
+    );
+  }
+
+  /**
+   *  Whether this scope is restricting the possible runtime types of the provided type.
+   */
+  isStrictlyRefining(type: GraphQLCompositeType) : boolean {
+    // This scope will refine the provided type, unless that provided type is a subtype of all
+    // the type in the chain.
+    let scope: Scope | undefined = this;
+    while (scope) {
+      if (scope.parentType !== type && isTypeSubTypeOf(this.context.schema, scope.parentType, type)) {
+        return true;
+      }
+      scope = scope.enclosing;
+    }
+    return false;
+  }
+
+  private computePossibleRuntimeTypes() : ReadonlyArray<GraphQLObjectType> {
+    // The possible runtime types is the intersection of all the possible types of each condition in scope.
+    let possibleTypes = this.context.getPossibleTypes(this.parentType);
+    let nextScope = this.enclosing;
+    while (nextScope) {
+      let enclosingPossibleTypes = this.context.getPossibleTypes(nextScope.parentType);
+      possibleTypes = possibleTypes.filter(t => enclosingPossibleTypes.includes(t));
+      nextScope = nextScope.enclosing;
+    }
+    return possibleTypes;
+  }
+
+  /**
+   * Computes the set of object types that an object within this scope can have at runtime.
+   */
+  possibleRuntimeTypes() : ReadonlyArray<GraphQLObjectType> {
+    if (!this.cachedRuntimeTypes) {
+      this.cachedRuntimeTypes = this.computePossibleRuntimeTypes();
+    }
+    return this.cachedRuntimeTypes;
+  }
+
+  private static valueIdentityKey(value: ValueNode) : string {
```

Comment:
Is encoding the type of the value really necessary here? Since these are used as directive arguments, the values should already be validated I think, so we can be sure their types match the definition?

Reply:
Probably not. I just felt this added a layer of security against weird cases I might not be thinking of, and this for not much extra effort. Happy to remove it if you feel this distract more than anything else (which might well be true).

Reply:
I'm not a fan of us having to come up with a custom encoding scheme if we can avoid it, because while it can protect against weird cases it could also be a source of bugs and may have to be tested separately. I'll leave the decision to you, because I don't have a strong opinion here. It seems to me the directive definitions should provide enough protection, but I could very well be missing something.